File: questnav-lib/src/main/java/gg/questnav/questnav/protos/wpilib/DeviceDataProto.java
/*
 * QUESTNAV
 * https://github.com/QuestNav
 * Copyright (C) 2025 QuestNav
 *
 * This file is part of QuestNav.
 *
 * QuestNav is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * QuestNav is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with QuestNav. If not, see <https://www.gnu.org/licenses/>.
 */

package gg.questnav.questnav.protos.wpilib;

import edu.wpi.first.util.protobuf.Protobuf;
import gg.questnav.questnav.protos.generated.Data;
import us.hebi.quickbuf.Descriptors;

/** WPILib Protobuf layer for DeviceData Protobuf */
public class DeviceDataProto
    implements Protobuf<Data.ProtobufQuestNavDeviceData, Data.ProtobufQuestNavDeviceData> {
  @Override
  public Class<Data.ProtobufQuestNavDeviceData> getTypeClass() {
    return Data.ProtobufQuestNavDeviceData.class;
  }

  @Override
  public Descriptors.Descriptor getDescriptor() {
    return Data.ProtobufQuestNavDeviceData.getDescriptor();
  }

  @Override
  public Data.ProtobufQuestNavDeviceData createMessage() {
    return Data.ProtobufQuestNavDeviceData.newInstance();
  }

  @Override
  public Data.ProtobufQuestNavDeviceData unpack(Data.ProtobufQuestNavDeviceData msg) {
    return msg.clone();
  }

  @Override
  public void pack(Data.ProtobufQuestNavDeviceData msg, Data.ProtobufQuestNavDeviceData value) {
    msg.copyFrom(value);
  }
}

File: questnav-lib/src/main/native/cpp/gg/questnav/questnav/protos/wpilib/DeviceDataProto.cpp
/*
* QUESTNAV
  https://github.com/QuestNav
* Copyright (C) 2025 QuestNav
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the MIT License as published.
*/
#include <wpi/protobuf/ProtobufCallbacks.h>

#include "gg/questnav/questnav/protos/wpilib/DeviceDataProto.h"

std::optional<questnav_protos_data_ProtobufQuestNavDeviceData >
wpi::Protobuf<questnav_protos_data_ProtobufQuestNavDeviceData >::Unpack(InputStream& stream) {
  questnav_protos_data_ProtobufQuestNavDeviceData  msg;
  if (!stream.Decode(msg)) {
    return std::nullopt;
  }
  return msg;
}

bool wpi::Protobuf<questnav_protos_data_ProtobufQuestNavDeviceData >::Pack(
    OutputStream& stream, const questnav_protos_data_ProtobufQuestNavDeviceData& value) {
  return stream.Encode(value);
}